### ISSUE TITLE
TypeScript does not infer equivalence of generic type arguments, TS2322 for equivalent types

### ISSUE THREAD

**Issue by thomaslaich** [microsoft/TypeScript#35679]
**TypeScript Version:** 3.7.2 or nightly

**Search Terms:** generic constrained type parameter error, error TS2322

**Code:** Consider the following example:
```typescript
function myFunc<TProps extends { foo: any }>(props: TProps) {
    type NewType = Omit<TProps & { state: any }, "state">;
    /* declare */ const nt: NewType = 0 as any;
    const test: TProps = props;
    const test2: TProps = nt; // error
}
```
**Actual behaviour:** This yields the following error message on the line where I try to assign `nt` to `test2`:
```
Type 'Pick<TProps & { state: any; }, Exclude<keyof TProps, "state">>' is not assignable to type 'TProps'.
  'Pick<TProps & { state: any; }, Exclude<keyof TProps, "state">>' is assignable to the constraint of type 'TProps', but 'TProps' could be instantiated with a different subtype of constraint '{ foo: any; }'.(2322)
```

**Expected behaviour:** Should have no error.

**Background:** This error message is correctly shown on something like this:
```typescript
function fn1<T extends string>(x: T): T {
  return "hello world!";
}
```
However, in the above example `NewType` should be the same as `TProps` no matter how `TProps` is instantiated.

**Playground Link:** [TS Playground](http://www.typescriptlang.org/play/index.html?ts=3.8.0-dev.20191212&ssl=6&ssc=2&pln=1&pc=1#code/GYVwdgxgLglg9mABAWwJ4DFwQDwBUAKATnAA4DOiApgB5SVgAmFA3osHHAFyICGYqiAL4A+ABQli5bgUlkAlImYAoRKsRRUJSogBylAO65N2gLyIA8shhQ8RUhQBkixGSg863PgMEAaRACJXd0p-YQBuFTUAegAqRAZKCAAbHkJtGKjECARXRDAobj1DY0QzAAZeCi8ItSycqHVKV2k7clLECXsatWywXLpXACYW2Xb8sMQozMpCYkIlQSA)

**Related issues:**
https://github.com/microsoft/TypeScript/issues/29049

**Comment by fatcerberus** [comment]
`TProps` might have a `state` property.  `NewType` definitely doesn’t (because you’ve `Omit`ted it).

**Comment by RyanCavanaugh** [comment]
Some day we will write an error message that convinces people these errors are correct 😢
